File: internal/agent/root/agents_setup.go
package root

import (
	"context"
	"fmt"
	"log/slog"

	"google.golang.org/adk/agent"
	"google.golang.org/adk/runner"

	"github.com/jkjamies/automation-agent/internal/agent/setup"
	"github.com/jkjamies/automation-agent/internal/ingest"
)

// Deps wires the dispatcher. SummaryAgent may be nil to leave the summary workflow
// disabled (e.g. when no repos are configured).
type Deps struct {
	SummaryAgent agent.Agent
	Log          *slog.Logger
}

// BuildRootDispatcher builds the dispatcher and registers the workflows that are
// available. The cron kinds route to the summary workflow.
func BuildRootDispatcher(d Deps) (*Dispatcher, error) {
	disp := NewDispatcher(d.Log)

	if d.SummaryAgent != nil {
		r, err := setup.NewRunner("automation-agent", d.SummaryAgent)
		if err != nil {
			return nil, fmt.Errorf("build summary runner: %w", err)
		}
		h := summaryHandler(r)
		disp.Register(ingest.KindCronDaily, h)
		disp.Register(ingest.KindCronWeekly, h)
	}

	// KindLint / KindCI are registered by the lint-fixer in a later phase.
	return disp, nil
}

// summaryHandler drives the summary workflow runner for a cron envelope, using a
// fresh session per fire.
func summaryHandler(r *runner.Runner) Handler {
	return func(ctx context.Context, e ingest.Envelope) error {
		sessionID := fmt.Sprintf("summary-%d", e.ReceivedAt.UnixNano())
		return setup.Drive(ctx, r, "system", sessionID, "Run the daily commit digest.")
	}
}

File: python/automation_agent/agent/root/agents_setup.py
"""Builds the root dispatcher and registers available workflows (port of
``root/agents_setup.go``).

Cron kinds → summary; LINT → lint-fixer; COVERAGE → coverage-fixer; CI → resume
(all fix engines). Each handler is optional.
"""

from __future__ import annotations

import logging
import time
from dataclasses import dataclass

from google.adk.agents import BaseAgent
from google.adk.runners import Runner

from automation_agent.agent import setup
from automation_agent.agent.root.root import Dispatcher, Handler
from automation_agent.ingest import Envelope, Kind


@dataclass
class Deps:
    """Wires the dispatcher. Each handler is optional.

    ``ci_resume`` handles :attr:`Kind.CI` for every fix workflow (lint, coverage) — each
    engine no-ops unless its check matches.
    """

    summary_agent: BaseAgent | None = None
    lint_kickoff: Handler | None = None  # Kind.LINT
    coverage_kickoff: Handler | None = None  # Kind.COVERAGE
    ci_resume: Handler | None = None  # Kind.CI (dispatched to all fix engines)
    log: logging.Logger | None = None


def build_root_dispatcher(d: Deps) -> Dispatcher:
    """Build the dispatcher and register the available workflows.

    Cron kinds → summary; LINT → lint-fixer; COVERAGE → coverage-fixer; CI → resume.
    """
    disp = Dispatcher(d.log)

    if d.summary_agent is not None:
        runner = setup.new_runner("automation-agent", d.summary_agent)
        handler = summary_handler(runner)
        disp.register(Kind.CRON_DAILY, handler)
        disp.register(Kind.CRON_WEEKLY, handler)
    if d.lint_kickoff is not None:
        disp.register(Kind.LINT, d.lint_kickoff)
    if d.coverage_kickoff is not None:
        disp.register(Kind.COVERAGE, d.coverage_kickoff)
    if d.ci_resume is not None:
        disp.register(Kind.CI, d.ci_resume)
    return disp


def summary_handler(runner: Runner) -> Handler:
    """Drive the summary workflow runner for a cron envelope, with a fresh session per
    fire."""

    async def handle(e: Envelope) -> None:
        session_id = f"summary-{time.monotonic_ns()}"
        await setup.drive(runner, "system", session_id, "Run the daily commit digest.")

    return handle
